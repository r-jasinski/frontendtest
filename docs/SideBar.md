# SideBar.vue

This Vue component represents the sidebar of the application. It is responsible for rendering the notations and scrolling the move list into view when a new move is made.

## Template Structure

The template structure is not included in the provided code snippet.

## Props

- `notations`: An array that contains the notations. It is required.

## Watchers

- `notations`: This watcher is triggered when a new move is made (i.e., when the `notations` prop changes). It calls the `scrollToElement` method to scroll the move list into view.

## Methods

- `scrollToElement()`: This method uses the `scrollIntoView` method of the `HTMLElement` interface to scroll the move list into view. The `behavior` option is set to 'smooth', so the scrolling is done smoothly rather than instantly.

## Styles

The sidebar has a height of 85vh, a width of 100%, and a background color of gray-900. It uses flexbox to align its children in a column and align them in the center. It also has a border radius of 8px, a font size of 12px, and a muted text color.
