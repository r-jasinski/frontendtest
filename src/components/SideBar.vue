<template>
  <div class="sidebar">
    <div class="tabs">
      <span>
        <TimerSVG />
        Play
      </span>
      <span>
        <MoreSVG />
        New Game
      </span>
      <span>
        <ChessBoardSVG />
        Games
      </span>
      <span>
        <PlayersSVG />
        Players
      </span>
    </div>
    <div class="opening">
      Center Game: Beyer Gambit
      <span><BookSVG /><CompassSVG /></span>
    </div>
    <div class="move-list">
      <ol>
        <li
          v-for="({ notation }, index) in notations"
          :key="`${index + 1}-${notation}`"
        >
          <span>{{ notation }}</span>
        </li>
        <li ref="moveListRef"></li>
      </ol>
    </div>
    <div class="game-buttons">
      <span><HalfSVG />Draw<FlagSVG />Resign</span>
      <span><FirstSVG /><PreviousSVG /><NextSVG /><LastSVG /><FlipSVG /></span>
    </div>
    <div class="chat">
      <span>NEW GAME</span>
      <span>Guest2939666985 vs Guest9241962816 (5 min Unrated)</span>
    </div>
    <div class="chat-login">
      Please <a href="#">Sign Up</a> to use the chat.
    </div>
  </div>
</template>

<script lang="ts">
import { defineComponent, PropType } from 'vue'
import ChessBoardSVG from '../assets/icons/board.svg?component'
import BookSVG from '../assets/icons/book.svg?component'
import CompassSVG from '../assets/icons/compass.svg?component'
import FirstSVG from '../assets/icons/first.svg?component'
import FlagSVG from '../assets/icons/flag.svg?component'
import FlipSVG from '../assets/icons/flip.svg?component'
import HalfSVG from '../assets/icons/half.svg?component'
import LastSVG from '../assets/icons/last.svg?component'
import MoreSVG from '../assets/icons/more.svg?component'
import NextSVG from '../assets/icons/next.svg?component'
import PlayersSVG from '../assets/icons/players.svg?component'
import PreviousSVG from '../assets/icons/previous.svg?component'
import TimerSVG from '../assets/icons/timer.svg?component'

import { Notation } from '../App.vue'

export default defineComponent({
  name: 'SideBar',
  components: {
    ChessBoardSVG,
    BookSVG,
    CompassSVG,
    FirstSVG,
    FlagSVG,
    FlipSVG,
    HalfSVG,
    LastSVG,
    MoreSVG,
    NextSVG,
    PlayersSVG,
    PreviousSVG,
    TimerSVG
  },
  props: {
    notations: Array as PropType<Notation[]>
  },
  watch: {
    notations: {
      deep: true,
      handler() {
        this.scrollToElement()
      }
    }
  },
  methods: {
    /**
     * Scrolls the move list into view.
     *
     * This method uses the `scrollIntoView` method of the `HTMLElement` interface to scroll the move list into view.
     * The `behavior` option is set to 'smooth', so the scrolling is done smoothly rather than instantly.
     */
    scrollToElement() {
      ;(this.$refs.moveListRef as HTMLElement).scrollIntoView({
        behavior: 'smooth'
      })
    }
  }
})
</script>

<style scoped>
.sidebar {
  height: 85vh;
  width: 100%;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: flex-start;
  border-radius: 8px;
  background-color: var(--color-gray-900);
  font-size: 12px;
  color: var(--color-text-muted);
}
* {
  scrollbar-width: thin;
  scrollbar-color: var(--color-gray-900) var(--color-bg-tertiary);
}
*::-webkit-scrollbar {
  width: 10px;
}
*::-webkit-scrollbar-track {
  background: var(--color-bg-tertiary);
  opacity: 0.4;
}
*::-webkit-scrollbar-thumb {
  background-color: var(--color-gray-900);
  border-radius: 10px;
  border: 2px solid var(--color-bg-tertiary);
  opacity: 0.4;
}
.tabs {
  width: 100%;
  height: 50px;
  display: flex;
}
.tabs span:first-of-type {
  color: var(--color-neutrals-white);
}
.tabs span {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  flex: 1;
  background-color: var(--color-bg-tertiary);
  /* color: var(--color-gray-300); */
}
.tabs span:first-of-type {
  background-color: transparent;
}
.opening {
  width: 100%;
  height: 42px;
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 0 12px;
  font-weight: 300;
  border-bottom: 1px solid var(--theme-background-color);
}
.opening span {
  display: flex;
  align-items: center;
  gap: 6px;
}
.move-list {
  width: 100%;
  height: 40%;
  display: flex;
  flex-direction: column;
  font-weight: 300;
  overflow-y: auto;
  overflow-x: hidden;
}
ol {
  list-style: none;
  padding: 0;
  margin: 0;
  counter-reset: item;
}
li {
  height: 24px;
  margin: 0 auto;
  padding-left: 32px;
  position: relative;
  line-height: 24px;
}
li:not(:last-of-type):nth-child(odd) {
  height: 24px;
  background-color: var(--color-bg-tertiary);
}
li:not(:last-of-type):before {
  content: counter(item) '\002E';
  counter-increment: item;
  position: absolute;
  translate: -16px;
  height: 24px;
}
li:last-of-type span {
  visibility: hidden;
  pointer-events: none;
}
li:nth-last-of-type(2) span {
  font-weight: 500;
  color: var(--color-neutrals-white);
}
li:nth-last-of-type(2) span:after {
  content: '';
  position: absolute;
  width: 42px;
  height: calc(100% - 4px);
  left: 28px;
  border-radius: 2px;
  background-color: var(--color-neutrals-white);
  border-bottom: 4px solid var(--color-bg-tertiary);
  opacity: 0.3;
}
.game-buttons {
  width: 100%;
  height: 42px;
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 0 12px;
  font-weight: 500;
  background-color: var(--color-bg-tertiary);
}
.game-buttons span {
  display: flex;
  gap: 8px;
}
.chat {
  width: 100%;
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  padding: 12px;
  font-weight: 300;
  flex-grow: 1;
}
.chat span:first-of-type {
  font-weight: 500;
}
.chat-login {
  width: 100%;
  height: 30px;
  display: flex;
  justify-content: flex-start;
  align-items: center;
  align-self: flex-end;
  padding: 0 12px;
  font-weight: 300;
  border-top: 1px solid var(--theme-background-color);
}
a {
  margin: 0 3px;
  color: var(--color-text-link);
  text-decoration: none;
}
</style>
