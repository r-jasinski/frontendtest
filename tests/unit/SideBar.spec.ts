import { describe, it, expect } from 'vitest'
import { mount } from '@vue/test-utils'
import SideBar from '../../src/components/SideBar.vue'

describe('SideBar.vue', () => {
  const wrapper = mount(SideBar, {
    propsData: {
      notations: [{ notation: 'a1', square: '8' }]
    }
  })
  it('should render notation', () => {
    expect(wrapper.find('li:nth-last-of-type(2) span').text()).toBe('a1')
  })
  it('should render icons', () => {
    expect(wrapper.findAll('svg')).toHaveLength(13)
  })
  it('should match snapshot', () => {
    expect(wrapper.element).toMatchSnapshot()
  })
})
