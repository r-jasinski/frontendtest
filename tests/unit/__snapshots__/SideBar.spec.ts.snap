// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SideBar.vue > should match snapshot 1`] = `
<div
  class="sidebar"
  data-v-ce868814=""
>
  <div
    class="tabs"
    data-v-ce868814=""
  >
    <span
      data-v-ce868814=""
    >
      <svg
        data-v-ce868814=""
        fill="none"
        height="21"
        width="18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10.264 3.18q3.282.45 5.509 2.97T18 12.03q0 3.72-2.634 6.345T9 21t-6.366-2.625T0 12.03q0-3.36 2.227-5.88 2.228-2.52 5.509-2.97v-.81H6.26V0h5.478v2.37h-1.475zM4.741 16.29Q6.5 18.06 9 18.06t4.274-1.77 1.776-4.26-1.776-4.245T9 6.03 4.74 7.785Q2.98 9.54 2.98 12.03t1.76 4.26m3.476-8.28h1.596l.782 4.89L9 14.43 7.405 12.9z"
          fill="currentColor"
        />
      </svg>
       Play 
    </span>
    <span
      data-v-ce868814=""
    >
      <svg
        data-v-ce868814=""
        fill="none"
        height="18"
        width="18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M14.596 0q.999 0 1.74.725.74.725.74 1.756v12.115q0 .999-.74 1.74-.741.74-1.74.74H2.48q-.999.001-1.74-.724A2.37 2.37 0 0 1 0 14.596V2.48q0-.999.741-1.74T2.481 0zm-1.063 9.505V7.572a.35.35 0 0 0-.097-.258.3.3 0 0 0-.226-.097H9.86v-3.35a.3.3 0 0 0-.097-.226.35.35 0 0 0-.258-.097H7.572a.35.35 0 0 0-.258.097.3.3 0 0 0-.097.225v3.351h-3.35a.3.3 0 0 0-.226.097.35.35 0 0 0-.097.258v1.933q0 .16.097.258a.3.3 0 0 0 .225.097h3.351v3.35q0 .13.097.226t.258.097h1.933a.35.35 0 0 0 .258-.097.3.3 0 0 0 .097-.226V9.86h3.35a.3.3 0 0 0 .226-.097.35.35 0 0 0 .097-.258"
          fill="currentColor"
          opacity=".7"
        />
      </svg>
       New Game 
    </span>
    <span
      data-v-ce868814=""
    >
      <svg
        data-v-ce868814=""
        fill="none"
        height="20"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19.29 19.29H0V0h19.29zm-.803-5.24V9.645h-4.405V5.208h4.405V.804h-4.405v4.404H9.645V.804H5.241v4.404H.804v4.437H5.24v4.405H.804v4.437H5.24V14.05h4.404v4.437h4.437V14.05zM5.24 9.645h4.404V5.208H5.241zm4.404 4.405h4.437V9.645H9.645z"
          fill="currentColor"
          opacity=".7"
        />
      </svg>
       Games 
    </span>
    <span
      data-v-ce868814=""
    >
      <svg
        data-v-ce868814=""
        fill="none"
        height="19"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          fill="currentColor"
          opacity=".7"
        >
          <path
            d="M20 15.062q0-4.75-1.426-5.261l-3.37-1.413q-.758-.391-.97-.902.728-.782.97-1.684l.213-.12q.395-1.053.334-2.044l-.152-.12q.091-1.444-.242-2.075-.364-.601-1.108-1.007Q13.506.03 12.807 0h-.364q-.698.03-1.441.436-.744.405-1.078 1.007-.365.6-.273 2.044.819.572 1.153 1.173.364.72.364 2.074l.091.09.03.361a5.7 5.7 0 0 1-.394 2.375l-.09.27-.183.121q-.213.571-.607 1.142.122.09.243.15l3.369 1.414q1.213.45 1.669 2.735H20z"
          />
          <path
            d="M14.75 18.67q0-4.75-1.427-5.262l-3.399-1.413q-.729-.39-.94-.902.728-.78.97-1.683l.183-.12q.425-1.083.333-2.045l-.151-.12q.15-1.383-.243-2.074-.334-.602-1.078-1.007-.743-.406-1.441-.436h-.364q-.73.03-1.472.436-.744.405-1.078 1.007-.333.57-.242 2.074l-.152.12q-.061.992.334 2.045l.212.12q.243.901.971 1.683-.273.572-.97.902l-3.37 1.413Q.001 13.92 0 18.67V19h14.75z"
          />
        </g>
      </svg>
       Players 
    </span>
  </div>
  <div
    class="opening"
    data-v-ce868814=""
  >
     Center Game: Beyer Gambit 
    <span
      data-v-ce868814=""
    >
      <svg
        data-v-ce868814=""
        fill="none"
        height="15"
        width="17"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12.067 15H2.674q-.713-.025-1.527-.5-.712-.575-.993-1.275-.28-.7-.05-1.275L4.251 1q.56-1 1.96-1h7.739q.79 0 1.196.55.433.5.178 1.25L12.5 10.675q-.178.6-.79 1.05-.585.425-1.221.425H1.376q-.33.2-.178.65.178.425.61.775.46.3.866.3h9.393a.8.8 0 0 0 .586-.225q.28-.225.381-.45l3.08-9.625.052-.325a.7.7 0 0 0-.026-.15v-.125q.356.175.611.425.382.5.178 1.275l-2.8 8.85q-.178.625-.764 1.05a2.16 2.16 0 0 1-1.298.425M4.89 8.475q0 .15.05.275.127.125.586.3.534.15 1.324.15.942-.1 1.323-.2.026.025.255-.025.23-.075.484-.225.255-.176.356-.375.077-.25-.305-.4.33-.275.33-.375l.077-.2q0-.225-.433-.375-.432-.125-.433-.275 0-.45.255-1.125l.102-.4Q8.936 5 8.962 4.9a.4.4 0 0 0 .05-.125q.739-.1.866-.35.05-.1-.05-.15-.078-.075-.306-.1l-.077-.05q.46-.375.611-.8.179-.7-.28-.925-.076-.1-.229-.075-.153 0-.484.075-.636.275-.916.875t.025.875a.4.4 0 0 0-.152.05q-.51.075-.586.3-.075.1.077.175.025 0 .101.025.076 0 .179.025.126.025.229.05Q7.307 6.399 6.976 6.8q-.178.125-.738.325-.687.225-.713.5-.101.225.127.4-.585.1-.763.45"
          fill="currentColor"
          opacity=".7"
        />
      </svg>
      <svg
        data-v-ce868814=""
        fill="none"
        height="16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M2.347 2.347Q4.693 0 8 0t5.653 2.347Q16 4.693 16 8t-2.347 5.653Q11.306 16 8 16q-3.307 0-5.653-2.347Q0 11.306 0 8q0-3.307 2.347-5.653m1.186 10.12Q5.387 14.32 8 14.32t4.467-1.853Q14.32 10.613 14.32 8t-1.853-4.467T8 1.68 3.533 3.533Q1.68 5.387 1.68 8t1.853 4.467m8.014-8.014-2.134 4.96-4.96 2.134 2.134-4.96zM5.28 10.72l3.573-1.547-2.026-2.026z"
          fill="currentColor"
          opacity=".7"
        />
      </svg>
    </span>
  </div>
  <div
    class="move-list"
    data-v-ce868814=""
  >
    <ol
      data-v-ce868814=""
    >
      
      <li
        data-v-ce868814=""
      >
        <span
          data-v-ce868814=""
        >
          a1
        </span>
      </li>
      
      <li
        data-v-ce868814=""
      />
    </ol>
  </div>
  <div
    class="game-buttons"
    data-v-ce868814=""
  >
    <span
      data-v-ce868814=""
    >
      <svg
        data-v-ce868814=""
        fill="none"
        height="14"
        width="18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M2.095 3.619V8.32h2.664V0H2.664q-.052.978-.828 1.413T0 1.85v1.77zM10.112 0 3 14h3l7.138-14zm.155 14H18v-2.087h-4.034q0-.131.62-.41.62-.276 1.345-.62.724-.343 1.345-1.07.62-.726.62-1.677 0-2.642-3.672-2.642-1.861 0-2.819.951-1.034.925-.957 2.695h2.405q-.025-.74.31-1.242.337-.501 1.061-.502a.97.97 0 0 1 .724.29q.285.291.285.767 0 .264-.22.515-.22.25-.4.37-.183.118-.544.304-.129.079-.75.435-.62.357-.815.489-.193.132-.685.489t-.685.607q-.195.252-.466.634a2 2 0 0 0-.349.806q-.078.423-.052.898"
          fill="currentColor"
          opacity=".7"
        />
      </svg>
      Draw
      <svg
        data-v-ce868814=""
        fill="none"
        height="15"
        width="15"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M14.325 4.747 15 7.273a4.38 4.38 0 0 1-2.125 1.262q-.8.202-2.1.202-.225 0-.638-.025-.411-.026-.75-.038a17 17 0 0 0-.637-.012q-1.3 0-2.1.202A4.2 4.2 0 0 0 4.525 10.1a1 1 0 0 0-.038-.19 4 4 0 0 0-.062-.214L3.85 7.576q-.075-.228-.263-.998t-.312-1.2q0-.024-.262-.997Q2.75 3.41 2.6 2.828q.85-.909 2.125-1.237.8-.202 2.075-.202.35 0 1.038.025.687.025 1.012.025 1.225.051 2.1-.202Q12.225.91 13.075 0l.675 2.525q.1.38.287 1.111.188.733.288 1.111M1.375 3.131 4.3 14.116a.66.66 0 0 1-.075.53.7.7 0 0 1-.425.329.64.64 0 0 1-.525-.076.7.7 0 0 1-.325-.43L.025 3.486a.66.66 0 0 1 .075-.53.7.7 0 0 1 .425-.329.64.64 0 0 1 .525.076.7.7 0 0 1 .325.43"
          fill="currentColor"
          opacity=".7"
        />
      </svg>
      Resign
    </span>
    <span
      data-v-ce868814=""
    >
      <svg
        data-v-ce868814=""
        fill="none"
        height="18"
        width="17"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M2.754.03q.504 0 .8.299.296.3.296.777v15.758q0 .509-.296.807-.296.3-.8.3L1.125 18q-.473 0-.8-.329A1.1 1.1 0 0 1 0 16.864V1.136Q0 .658.326.33t.8-.329zm8.027 9 5.893 5.83q.326.39.326.778t-.326.777l-1.155 1.196q-.385.33-.8.33-.413 0-.8-.33L6.19 9.748a1.03 1.03 0 0 1-.355-.778q0-.418.356-.777L13.92.329q.385-.33.8-.329.413 0 .8.329l1.154 1.166q.326.39.326.807 0 .449-.326.778z"
          fill="currentColor"
          opacity=".7"
        />
      </svg>
      <svg
        data-v-ce868814=""
        fill="none"
        height="18"
        width="11"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m.322 9.78 7.665 7.89q.321.33.76.33.41 0 .79-.33l1.141-1.2q.322-.39.322-.78t-.322-.78L4.886 9.06l5.792-5.97Q11 2.76 11 2.31q0-.42-.322-.81L9.538.33Q9.155 0 8.747 0q-.44 0-.761.33L.322 8.22Q0 8.61 0 9q0 .45.322.78"
          fill="currentColor"
          opacity=".7"
        />
      </svg>
      <svg
        data-v-ce868814=""
        fill="none"
        height="19"
        width="11"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m10.678 10.323-7.665 8.329a1 1 0 0 1-.76.348q-.41 0-.79-.348L.322 17.385Q0 16.973 0 16.562q0-.413.322-.824l5.792-6.175L.322 3.262A1.17 1.17 0 0 1 0 2.438q0-.443.322-.855L1.462.348Q1.844 0 2.253 0q.44 0 .761.348l7.665 8.329Q11 9.088 11 9.5q0 .475-.322.823"
          fill="currentColor"
          opacity=".7"
        />
      </svg>
      <svg
        data-v-ce868814=""
        fill="none"
        height="19"
        width="18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m3.261.347 8.185 8.3q.376.38.376.821 0 .505-.376.821l-8.185 8.3q-.407.348-.846.348-.44 0-.847-.347L.345 17.327Q0 16.917 0 16.507t.345-.821l6.24-6.154L.345 3.25A1.13 1.13 0 0 1 0 2.43q0-.443.345-.853L1.568.348Q1.976 0 2.415 0t.846.347M16.808 0q.502 0 .847.347Q18 .695 18 1.2v16.602q0 .504-.345.852-.345.347-.847.347l-1.724-.032q-.534 0-.847-.315-.314-.316-.314-.852V1.168q0-.505.314-.82.313-.316.847-.316z"
          fill="currentColor"
          opacity=".7"
        />
      </svg>
      <svg
        data-v-ce868814=""
        fill="none"
        height="19"
        width="22"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M9.315 14.744q.073.076.147.228a1 1 0 0 0 .22.228l2.856 3.344H4.59q-.366 0-.66-.304a.9.9 0 0 1-.256-.646V6.65H.817q-.622 0-.769-.38-.183-.343.33-.874L4.773.342Q5.103 0 5.506 0q.44 0 .732.342l4.396 5.054q.512.532.33.874-.11.38-.77.38H7.337v8.094zm11.868-2.394q.622 0 .769.38.183.343-.33.874l-4.395 5.054a.93.93 0 0 1-.733.342.93.93 0 0 1-.732-.342l-4.396-5.054q-.512-.532-.33-.874.11-.38.77-.38h2.857V4.294h-1.978a.9.9 0 0 1-.147-.228 3 3 0 0 0-.183-.266L9.462.494h7.948q.402 0 .66.266.256.342.256.684V12.35z"
          fill="currentColor"
          opacity=".7"
        />
      </svg>
    </span>
  </div>
  <div
    class="chat"
    data-v-ce868814=""
  >
    <span
      data-v-ce868814=""
    >
      NEW GAME
    </span>
    <span
      data-v-ce868814=""
    >
      Guest2939666985 vs Guest9241962816 (5 min Unrated)
    </span>
  </div>
  <div
    class="chat-login"
    data-v-ce868814=""
  >
     Please 
    <a
      data-v-ce868814=""
      href="#"
    >
      Sign Up
    </a>
     to use the chat. 
  </div>
</div>
`;
